File: 3244/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(shortestDistanceAfterQueries(5, [][]int{{2,4},{0,2},{0,4}}))
}

func shortestDistanceAfterQueries(n int, queries [][]int) []int {
	roads := make([]int, n)
	for i := 0; i < n; i++ {
		roads[i] = i + 1
	}
	var res []int
	dist := n-1
	for _, query := range queries {
		k := roads[query[0]]
		roads[query[0]] = query[1]
		for k != -1 && k < query[1] {
			k, roads[k] = roads[k], -1
			dist--
		}
		res = append(res, dist)
	}
	return res
}

File: 3244/main.rs
struct Solution{}

impl Solution {
    pub fn shortest_distance_after_queries(n: i32, queries: Vec<Vec<i32>>) -> Vec<i32> {
        let mut roads: Vec<i32> = (1..=n).collect();
        let mut res: Vec<i32> = Vec::new();
        let mut dist = n-1;
        for query in &queries {
            let mut k = roads[query[0] as usize];
            roads[query[0] as usize] = query[1];
            while k != -1 && k < query[1] {
                let t = roads[k as usize];
                roads[k as usize] = -1;
                k = t;
                dist -= 1;
            }
            res.push(dist);
        }
        res
    }
}

fn main() {
    println!("{:#?}", Solution::shortest_distance_after_queries(5, vec![vec![2,4], vec![0,2], vec![0,4]]));
}